geode-wan/src/upgradeTest/java/org/apache/geode/cache/wan/WANRollingUpgradeNewSenderProcessOldEvent.java

```diff
@@ -0,0 +1,411 @@
+        site2Locators);
+
+    // Roll mixed site servers to current
+    rollStartAndConfigureServerToCurrent(site1Server1, site1Locators, site2DistributedSystemId,
+        regionName, site1SenderId, ParallelGatewaySenderQueue.DEFAULT_MESSAGE_SYNC_INTERVAL);
+    site1Server1.invoke(() -> pauseSender(site1SenderId));
+
+    // wait until site1Server finished rolling
+    site1Server1.invoke(() -> waitForQueueRegionToCertainSize(site1SenderId, numPuts, false));
+
+    rollStartAndConfigureServerToCurrent(site1Server2, site1Locators, site2DistributedSystemId,
+        regionName, site1SenderId, ParallelGatewaySenderQueue.DEFAULT_MESSAGE_SYNC_INTERVAL);
+    site1Server2.invoke(() -> pauseSender(site1SenderId));
+
+    // double check queue size should be numPuts
+    site1Server1.invoke(() -> waitForQueueRegionToCertainSize(site1SenderId, numPuts, false));
+    site1Server2.invoke(() -> waitForQueueRegionToCertainSize(site1SenderId, numPuts, false));
+
+    // Start and configure old site servers
+    String site2SenderId = getName() + "_gatewaysender_" + site1DistributedSystemId;
+    startAndConfigureServers(site2Server1, site2Server2, site2Locators, site1DistributedSystemId,
+        regionName, site2SenderId, ParallelGatewaySenderQueue.DEFAULT_MESSAGE_SYNC_INTERVAL);
+
+    // resume the senders at mixed site
+    site1Server1.invoke(() -> resumeSender(site1SenderId));
+    site1Server2.invoke(() -> resumeSender(site1SenderId));
+
+    // Wait for mixed site queues to be empty
+    site1Server1.invoke(() -> waitForQueueRegionToCertainSize(site1SenderId, 0, false));
+    site1Server2.invoke(() -> waitForQueueRegionToCertainSize(site1SenderId, 0, false));
+
+    // Verify remote site received events
+    int remoteServer1EventsReceived = site2Server1.invoke(() -> getEventsReceived(regionName));
+    int remoteServer2EventsReceived = site2Server2.invoke(() -> getEventsReceived(regionName));
+    assertEquals(numPuts, remoteServer1EventsReceived + remoteServer2EventsReceived);
+
+    // Clear events received in both sites
+    site1Server1.invoke(() -> clearEventsReceived(regionName));
+    site1Server2.invoke(() -> clearEventsReceived(regionName));
+    site2Server1.invoke(() -> clearEventsReceived(regionName));
+    site2Server2.invoke(() -> clearEventsReceived(regionName));
+  }
+
+  /**
+   * get new events to flow back to the old member and have the old member try to deserialize
+   * the new event
+   * 1.create 2 old senders, pause
+   * 2.do puts, build up the queue.
+   * 3.roll one sender to current
+   * 4.do additional puts that get queued in the new member with a redundant copy on the old member
+   * 5.remove the new member
+   * 6.start sender on old member
+   *
+   * The old sender should handle both old and new events, the remove site should receive all of
+   * them
+   */
+  @Test
+  public void bothOldAndNewEventsShouldBeProcessedByOldSender() {
```

Comment:
Can this be broken up until 6 smaller functions to make this overall function easier to read given it's already over 100 lines.

Reply:
I realize that these are test cases so maybe you can ignore this request.